// @chaingpt/ainews/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Errors = exports.AINews = void 0;
const axios_1 = __importDefault(require("axios"));
const error_1 = require("./error");
const config_1 = require("./config");
class AINews {
    constructor(obj) {
        this.apiKey = obj.apiKey;
    }
    getNews(findNewsDto) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            let obj = findNewsDto;
            Object.keys(findNewsDto).forEach((key) => {
                const typedKey = key;
                if (findNewsDto[typedKey] === undefined) {
                    delete obj[typedKey];
                }
            });
            const instance = axios_1.default.create({
                baseURL: config_1.SERVER_URL,
                timeout: config_1.TIMEOUT,
                headers: { 'Authorization': `Bearer ${this.apiKey}` },
                params: obj
            });
            try {
                const resp = yield instance.get(`/news`);
                return resp.data;
            }
            catch (error) {
                if (axios_1.default.isAxiosError(error)) {
                    switch ((_a = error.response) === null || _a === void 0 ? void 0 : _a.status) {
                        case 403:
                            throw new error_1.InvalidApiKeyError();
                        case 429:
                            throw new error_1.RateLimitExceededError();
                    }
                }
                throw new error_1.AINewsError((_c = (_b = error.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.message);
            }
        });
    }
}
exports.AINews = AINews;
exports.Errors = {
    AINewsError: error_1.AINewsError,
    InvalidApiKeyError: error_1.InvalidApiKeyError,
    RateLimitExceededError: error_1.RateLimitExceededError
};
exports.default = { AINews, Errors: exports.Errors };
